src/rt/cover.c

```diff
@@ -776,6 +814,186 @@ void x_cover_setup_toggle_cb(sig_shared_t *ss, int32_t *toggle_mask)
+}
+
+void cover_exclude_hier(cover_tagging_t *tagging, cover_exclude_ctx_t *ctx,
+                        const char *excl_hier, const char *bin)
+{
+   bool match = false;
+   int len = strlen(excl_hier);
+
+   for (int i = 0; i < tagging->tags.count; i++) {
+      cover_tag_t *tag = AREF(tagging->tags, i);
+
+      if (tag->kind == TAG_HIER || tag->kind == TAG_LAST)
+         continue;
+
+      if (ident_glob(tag->hier, excl_hier, len)) {
+         match = true;
+
+         // By default, exclude all bins of a tag
+         uint32_t bmask = (tag->kind == TAG_STMT) ?
+                              0xFFFFFFFF : (tag->flags & (COVER_FLAGS_ALL_BINS));
+
+#ifdef COVER_DEBUG
+         printf("Applying matching exclude:\n");
+         printf("    Tag:        %s\n", istr(tag->hier));
+         printf("    Tag flags:  %x\n", tag->flags);
+         printf("    Tag data:   %x\n", tag->data);
+         printf("    Bin mask:   %x\n", bmask);
+#endif
+
+         if (bin) {
+            if (tag->kind == TAG_STMT)
+               cover_exclude_error(ctx, "Statements do not contain bins,"
+                                   "but bin '%s' was given for statement: '%s'",
+                                   bin, istr(tag->hier));
+            bmask = cover_bin_str_to_bmask(ctx, bin, tag);
+         }
+
+         // Info about exclude
+         if (tag->kind == TAG_STMT)
+            cover_exclude_info(ctx, "Excluding statement: '%s'", istr(tag->hier));
+         else {
+            char LOCAL *blist = cover_bmask_to_bin_list(bmask);
+            cover_exclude_info(ctx, "Excluding %s: '%s' bins: %s",
+                                     tag_kind_str[tag->kind], istr(tag->hier),
+                                     blist);
+         }
+
+         // Check attempt to exclude something already covered
+         uint32_t excl_cov = bmask & tag->data;
+         if (excl_cov) {
+            if (tag->kind == TAG_STMT)
+               cover_exclude_warning(ctx, "Statement: '%s' already covered!",
+                                     istr(tag->hier));
+            else {
+               char LOCAL *blist = cover_bmask_to_bin_list(excl_cov);
+               cover_exclude_warning(ctx, "%s: '%s' bins: %s already covered!",
+                                     tag_kind_str[tag->kind], istr(tag->hier),
+                                     blist);
+            }
+         }
+
+         // Mark as excluded
+         tag->excl_msk |= bmask;
+      }
+   }
+
+   if (!match)
+      cover_exclude_warning(ctx,
+         "Exluded hierarchy does not match any coverage item: '%s'", excl_hier);
+}
+
+void cover_load_exclude_file(const char *path, cover_tagging_t *tagging)
+{
+   cover_exclude_ctx_t ctx = {0};
+
+   ctx.ef = fopen(path, "r");
+   if (ctx.ef == NULL)
+      fatal("Failed to open exclude file: %s\n", path);
+
+   char *delim = " ";
+   ssize_t read;
+
+   while ((read = getline(&(ctx.line), &(ctx.line_len), ctx.ef)) != -1) {
+
+      // Strip comments and newline
```

Comment:
I'd do this directly with a for loop:

```c
char *p = ctx.line;
for (; *p != '\0' && *p != '#' && *p != '\n');
*p = '\0;
```

It's less code and only makes one pass over the array.

Reply:
Or with a library function you could use `strpbrk`:

```c
char *stop = strpbrk(ctx.line, "#\n");
if (stop)
   *stop = '\0';
```